packages/mui-base/src/ListboxUnstyled/defaultListboxReducer.ts

```diff
@@ -0,0 +1,315 @@
+        ...state,
+        highlightedIndex: moveHighlight(-1, 'previous', !(disableListWrap ?? false)),
+      };
+
+    case 'ArrowDown':
+      // TODO: extend current selection with Shift modifier
+      return {
+        ...state,
+        highlightedIndex: moveHighlight(1, 'next', !(disableListWrap ?? false)),
+      };
+
+    case 'Enter':
+    case ' ':
+      if (state.highlightedIndex !== -1) {
+        const option = options[state.highlightedIndex];
+        const disabled = isOptionDisabled
+          ? isOptionDisabled(option, state.highlightedIndex)
+          : false;
+
+        if (disabled) {
+          return state;
+        }
+
+        if (selectMultiple) {
+          if (((state.selectedValue as TOption[]) ?? []).some((o) => optionComparer(o, option))) {
+            return {
+              ...state,
+              selectedValue: (state.selectedValue as TOption[]).filter(
+                (o) => !optionComparer(o, option),
+              ),
+            };
+          }
+
+          return {
+            ...state,
+            selectedValue: [...((state.selectedValue as TOption[]) ?? []), option],
+          };
+        }
+
+        return {
+          ...state,
+          selectedValue: option,
+        };
+      }
+      break;
+
+    default:
+      break;
+  }
+
+  return state;
+}
+
+function handleOptionClick<TOption>(
+  option: TOption,
+  state: ListboxState<TOption>,
+  props: UseListboxStrictProps<TOption>,
+): ListboxState<TOption> {
+  const {
+    multiple: selectMultiple,
+    optionComparer = (o, v) => o === v,
+    isOptionDisabled = () => false,
+  } = props;
+  const { selectedValue } = state;
+
+  const optionIndex = props.options.indexOf(option);
+
+  if (isOptionDisabled(option, optionIndex)) {
+    return state;
+  }
+
+  if (selectMultiple) {
+    const newSelectedValue = ((selectedValue as TOption[]) ?? []).some((v) =>
+      optionComparer(v, option),
+    )
+      ? (selectedValue as TOption[]).filter((v) => !optionComparer(v, option))
+      : [...((selectedValue as TOption[]) ?? []), option];
+    return {
+      ...state,
+      selectedValue: newSelectedValue,
+      highlightedIndex: optionIndex,
+    };
+  }
+
+  if (selectedValue != null && optionComparer(option, selectedValue as TOption)) {
+    return state;
+  }
+
+  return {
+    ...state,
+    selectedValue: option,
+    highlightedIndex: optionIndex,
+  };
+}
+
+function handleBlur<TOption>(state: ListboxState<TOption>): ListboxState<TOption> {
+  return {
+    ...state,
+    highlightedIndex: -1,
+  };
+}
+
+function handleOptionsChange<TOption>(
+  options: TOption[],
+  previousOptions: TOption[],
+  state: ListboxState<TOption>,
+  props: UseListboxStrictProps<TOption>,
+): ListboxState<TOption> {
+  const highlightedOption = previousOptions[state.highlightedIndex];
+  const hightlightedOptionNewIndex = options.findIndex((option) =>
+    props.optionComparer(option, highlightedOption),
+  );
+
+  if (props.multiple) {
```

Comment:
This too looks like a duplication, would be great if we can extract one method and use it in all three places.